Source/dataframe.lisp

```diff
@@ -0,0 +1,245 @@
+    (typecase row
+      (integer (setf row (elt (rows df) row)))
+      ;; Undocumented hack to map over col names.
+      (null (setf row (make-array (length (rows df))))))
+    (let ((cols (cols df)))
+      (flet ((norm-index (col)
+               (if (integerp col)
+                   col
+                   (position col cols :test #'string=))))
+        (etypecase col-slice
+          ((eql t)
+           (loop for index from 0
+                 for name across cols
+                 for value across row
+                 do (funcall fun index name value)))
+          ((cons (eql t) %col-index%)
+           (loop with i2 = (norm-index (cdr col-slice))
+                 for index from 0 upto i2
+                 for name across cols
+                 for value across row
+                 do (funcall fun index name value)))
+          ((cons %col-index% (eql t))
+           (loop with i1 = (norm-index (car col-slice))
+                 for index from i1
+                 for name = (elt cols index)
+                 for value = (elt row index)
+                 do (funcall fun index name value)))
+          ((cons %col-index% %col-index%)
+           (loop with i1 = (norm-index (car col-slice))
+                 with i2 = (norm-index (cdr col-slice))
+                 for index from i1 upto i2
+                 for name = (elt cols index)
+                 for value = (elt row index)
+                 do (funcall fun index name value)))
+          ((vector %col-index%)
+           (loop for col-index across col-slice
+                 for index = (norm-index col-index)
+                 for name = (elt cols index)
+                 for value = (elt row index)
+                 do (funcall fun index name value))))))))
+
+(defgeneric add-rows! (<data-frame> &rest rows)
+  (:method ((df <raw-data-frame>) &rest rows)
+    (loop with len = (length (cols df))
+          for row in rows
+          for new-row = (copy-array (coerce row 'vector) :adjustable t)
+          do (assert (sequence-of-length-p new-row len)
+                     nil
+                     '<row-length-mismatch>)
+             (vector-push-extend new-row (rows df)))))
+
+(defgeneric add-cols! (<data-frame> &rest name-fun-pairs)
+  (:method ((df <raw-data-frame>) &rest name-fun-pairs)
+    (loop for (name fun) in name-fun-pairs by #'cddr
+          do (assert (not (find name (cols df) :test #'string=))
+                     nil
+                     '<col-name-not-unique>)
+             (vector-push-extend name (cols df))
+             (loop for ind from 0
+                   for row in (rows df)
+                   for val = (funcall fun ind row)
+                   do (vector-push-extend val row)))))
+
+(defun make-data-frame (cols &rest rows)
+  (let* ((new-cols (copy-array (coerce cols 'vector) :adjustable t))
+         (new-rows (make-array (length rows) :adjustable t))
+         (col-width (length new-cols))
+         (index 0))
+    (map nil (lambda (row)
```

Comment:
Would it be worth checking the columns here too?
`
    (assert (not (= col-width (length (remove-duplicates cols))))
            nil 
            '<col-name-not-unique>)
`

Reply:
I think you mean

```
(assert (= col-width (length (remove-duplicates cols :test #'string=)))
            nil
            '<col-name-not-unique>)
```

yes, good idea. fixed.